### ISSUE TITLE
Error: /undefined in getenv

### ISSUE THREAD

**Issue by piyushrpt** [doxygen/doxygen#7484]
I was observing the same errors as ones reported in #7290 on my setup.

Macports on OSX 10.14.6
gs - 9.50
doxygen - 1.8.16 

On update to 1.8.17 via Macports - I see this new error:

```
Generating image Error: /undefined in getenv
Operand stack:
   (outfile)
Execution stack:
   %interp_exit   .runexec2   --nostringval--   --nostringval--   --nostringval--   2   %stopped_push   --nostringval--   --nostringval--   --nostringval--   false   1   %stopped_push   1990   1   4   %oparray_pop   1989   1   4   %oparray_pop   1977   1   4   %oparray_pop   1833   1   4   %oparray_pop   --nostringval--   %errorexec_pop   .runexec2   --nostringval--   --nostringval--   --nostringval--   2   %stopped_push   --nostringval--
Dictionary stack:
   --dict:731/1123(ro)(G)--   --dict:0/20(G)--   --dict:76/200(L)--   --dict:5/30(L)--   --dict:11/30(L)--
Current allocation mode is local
Current file position is 8386
GPL Ghostscript 9.50: Unrecoverable error, exit code 1
```

**Comment by albert-github** [comment]
Without a small example (including the output directories) it is hard to tel what is wrong.
- Can you please attach a, small, self contained example (source+configuration file in a tar or zip) that allows us to reproduce the problem?  Please don't add external links as they might not be persistent.

also the output generated when running `doxygen -d extcmd` is of interest.

Do you have the possibility to run with gs 9.27 and report from this version also the output and the result when running `doxygen -d extcmd`.

**Comment by piyushrpt** [comment]
I will try to create a small stand alone example. I dont have the setup to try gs 9.27. Here is the output with "-d extcmd" for now

```
Generating image form_61.png for formula
Executing external command `dvips -q -D 600 -n 1 -p 62 -o _form61_tmp.ps _formulas.dvi`
Executing external command `ps2epsi _form61_tmp.ps _form61.eps`
Error: /undefined in getenv
Operand stack:
   (outfile)
Execution stack:
   %interp_exit   .runexec2   --nostringval--   --nostringval--   --nostringval--   2   %stopped_push   --nostringval--   --nostringval--   --nostringval--   false   1   %stopped_push   1990   1   4   %oparray_pop   1989   1   4   %oparray_pop   1977   1   4   %oparray_pop   1833   1   4   %oparray_pop   --nostringval--   %errorexec_pop   .runexec2   --nostringval--   --nostringval--   --nostringval--   2   %stopped_push   --nostringval--
Dictionary stack:
   --dict:731/1123(ro)(G)--   --dict:0/20(G)--   --dict:76/200(L)--   --dict:5/30(L)--   --dict:11/30(L)--
Current allocation mode is local
Current file position is 8386
GPL Ghostscript 9.50: Unrecoverable error, exit code 1
```

**Comment by albert-github** [comment]
Maybe already forward the `_form61...` as well as the `_formulas...` files or have a look in them so you know which formula is causing the problem.

**Comment by piyushrpt** [comment]
Here is a simple index.dox that reproduces the issue. I used the default doxygen file generated with "doxygen -g" with this file. 

The original documented had 92 formulas that the "getenv" error is generated for each of them. 
```
/*!
\mainpage Formula Formula Testing


\section sec1 Testing 

This is an example formula

\f[ 
    A = B + C + \frac{D}{E}
\f]

\f$ P = \frac{Q}{R} \f$

*/
```

The corresponding formula.repository file that was generated was
```
\_form#0:\[ A = B + C + \frac{D}{E} \]
\_form#1:$ P = \frac{Q}{R} $
```

Here is the end of the screen output for this example 
```
No file _formulas.aux.
[1] [2] (./_formulas.aux) )
Output written on _formulas.dvi (2 pages, 532 bytes).
Transcript written on _formulas.log.
Generating image form_0.png for formula
Error: /undefined in getenv
Operand stack:
   (outfile)
Execution stack:
   %interp_exit   .runexec2   --nostringval--   --nostringval--   --nostringval--   2   %stopped_push   --nostringval--   --nostringval--   --nostringval--   false   1   %stopped_push   1990   1   4   %oparray_pop   1989   1   4   %oparray_pop   1977   1   4   %oparray_pop   1833   1   4   %oparray_pop   --nostringval--   %errorexec_pop   .runexec2   --nostringval--   --nostringval--   --nostringval--   2   %stopped_push   --nostringval--
Dictionary stack:
   --dict:731/1123(ro)(G)--   --dict:0/20(G)--   --dict:76/200(L)--   --dict:5/30(L)--   --dict:11/30(L)--
Current allocation mode is local
Current file position is 8386
GPL Ghostscript 9.50: Unrecoverable error, exit code 1
Generating image form_1.png for formula
Error: /undefined in getenv
Operand stack:
   (outfile)
Execution stack:
   %interp_exit   .runexec2   --nostringval--   --nostringval--   --nostringval--   2   %stopped_push   --nostringval--   --nostringval--   --nostringval--   false   1   %stopped_push   1990   1   4   %oparray_pop   1989   1   4   %oparray_pop   1977   1   4   %oparray_pop   1833   1   4   %oparray_pop   --nostringval--   %errorexec_pop   .runexec2   --nostringval--   --nostringval--   --nostringval--   2   %stopped_push   --nostringval--
Dictionary stack:
   --dict:731/1123(ro)(G)--   --dict:0/20(G)--   --dict:76/200(L)--   --dict:5/30(L)--   --dict:11/30(L)--
Current allocation mode is local
Current file position is 8386
GPL Ghostscript 9.50: Unrecoverable error, exit code 1
Generating index page...
Generating page index...
Generating module index...
Generating namespace index...
Generating namespace member index...
Generating annotated compound index...
Generating alphabetical compound index...
Generating hierarchical class index...
Generating member index...
Generating file index...
Generating file member index...
Generating example index...
finalizing index lists...
writing tag file...
Running plantuml with JAVA...
lookup cache used 0/65536 hits=0 misses=0
finished...
```

**Comment by albert-github** [comment]
Are there still `_formulas` or `formulas` files left on the html directory (so I can inspect them)?

As written before (also including the output directories):
- Can you please attach a, small, self contained example (source+configuration file in a tar or zip) that allows us to reproduce the problem? Please don't add external links as they might not be persistent.

I don't have gs 9.50 installed (yet, new system), but maybe I do see something in the output already.

**Comment by piyushrpt** [comment]
Here is the example folder with output. There are no other _formula files in the html folder. 
[doxy_getenv_gs950.zip](https://github.com/doxygen/doxygen/files/4027423/doxy_getenv_gs950.zip)

**Comment by albert-github** [comment]
All intermediate files are removed by doxygen ( :-( ), I will have to install gs 9.50 and see what happens.
It might be a special situation on Mac that is not reproducible on Windows, but I'll have a go. 

Seen the error message is there anything special in your environment?

Are you aware of MathJax (USE_MATHJAX=YES)?

**Comment by piyushrpt** [comment]
Yes. USE_MATHJAX works fine. 

This is a potential solution - I probably need to update some of the config files when the docs get published to gh-pages. We use sphinx + doxygen and I will look into updating both to use mathjax consistently. 

I dont think there is anything special with my env - I get everything via macports and didn't see issues before switching to gs 9.50.

**Comment by albert-github** [comment]
MathJax won't give a problem during compilation ass it will render the formulas when the HTML page is viewed.  Quality of the MathJax formulas is better than the png versions, so when possible it is advised to use MathJax for the HTML documentation.

Nonetheless the mentioned problem has to be checked as well.

I did with my current installation a small test and generated the intermediate files so we can run the problem outside of doxygen. I placed the required files and commands in an attachment (
[forms.tar.gz](https://github.com/doxygen/doxygen/files/4030148/forms.tar.gz) ). You can unpack the archive and run the file `cmds` (please have a look in it to see whether or not I'm doing some mean things 😉 ). Please send me all files from the directory in compressed tar filer (`tar czf result_forms.tar.gz _* cmds`) back.

**Comment by IsaacPascual** [comment]
Same problem with doxygen 1.8.17 and ghostscript 9.50.
I was waiting cause #7392 but the release don't help to fix ghostscript usage.

**Comment by albert-github** [comment]
@IsaacPascual 
- Which OS did you use?
- did you have a go with the forms.tar.gz I attached in my previous comment?

**Comment by IsaacPascual** [comment]
Archlinux

**Comment by IsaacPascual** [comment]
I try [forms.zip](https://github.com/doxygen/doxygen/files/4030472/forms.zip)
the output:
```
latex _formulas.tex
This is pdfTeX, Version 3.14159265-2.6-1.40.20 (TeX Live 2019/Arch Linux) (preloaded format=latex)
 restricted \write18 enabled.
entering extended mode
(./_formulas.tex
LaTeX2e <2019-10-01> patch level 1
(/usr/share/texmf-dist/tex/latex/base/article.cls
Document Class: article 2019/08/27 v1.4j Standard LaTeX document class
(/usr/share/texmf-dist/tex/latex/base/size10.clo))
(/usr/share/texmf-dist/tex/latex/base/ifthen.sty)
(/usr/share/texmf-dist/tex/latex/graphics/epsfig.sty
(/usr/share/texmf-dist/tex/latex/graphics/graphicx.sty
(/usr/share/texmf-dist/tex/latex/graphics/keyval.sty)
(/usr/share/texmf-dist/tex/latex/graphics/graphics.sty
(/usr/share/texmf-dist/tex/latex/graphics/trig.sty)
(/usr/share/texmf-dist/tex/latex/graphics-cfg/graphics.cfg)
(/usr/share/texmf-dist/tex/latex/graphics-def/dvips.def))))
(/usr/share/texmf-dist/tex/latex/base/inputenc.sty)
(/usr/share/texmf-dist/tex/latex/newunicodechar/newunicodechar.sty)
No file _formulas.aux.
[1] [2] (./_formulas.aux) )
Output written on _formulas.dvi (2 pages, 532 bytes).
Transcript written on _formulas.log.
=================================
dvips -q -D 600 -n 1 -p 1 -o _form0_tmp.ps _formulas.dvi
=================================
ps2epsi _form0_tmp.ps _form0.eps
Error: /undefined in getenv
Operand stack:
   (outfile)
Execution stack:
   %interp_exit   .runexec2   --nostringval--   --nostringval--   --nostringval--   2   %stopped_push   --nostringval--   --nostringval--   --nostringval--   false   1   %stopped_push   1990   1   4   %oparray_pop   1989   1   4   %oparray_pop   1977   1   4   %oparray_pop   1833   1   4   %oparray_pop   --nostringval--   %errorexec_pop   .runexec2   --nostringval--   --nostringval--   --nostringval--   2   %stopped_push   --nostringval--
Dictionary stack:
   --dict:731/1123(ro)(G)--   --dict:0/20(G)--   --dict:76/200(L)--   --dict:5/30(L)--   --dict:11/30(L)--
Current allocation mode is local
Current file position is 8386
GPL Ghostscript 9.50: Unrecoverable error, exit code 1
=================================
gs -q -g386x120 -r384x384 -sDEVICE=ppmraw -sOutputFile=_form0.pnm -dNOPAUSE -dBATCH -dNOSAFER _form0.ps
=================================
dvips -q -D 600 -n 1 -p 2 -o _form1_tmp.ps _formulas.dvi
=================================
ps2epsi _form1_tmp.ps _form1.eps
Error: /undefined in getenv
Operand stack:
   (outfile)
Execution stack:
   %interp_exit   .runexec2   --nostringval--   --nostringval--   --nostringval--   2   %stopped_push   --nostringval--   --nostringval--   --nostringval--   false   1   %stopped_push   1990   1   4   %oparray_pop   1989   1   4   %oparray_pop   1977   1   4   %oparray_pop   1833   1   4   %oparray_pop   --nostringval--   %errorexec_pop   .runexec2   --nostringval--   --nostringval--   --nostringval--   2   %stopped_push   --nostringval--
Dictionary stack:
   --dict:731/1123(ro)(G)--   --dict:0/20(G)--   --dict:76/200(L)--   --dict:5/30(L)--   --dict:11/30(L)--
Current allocation mode is local
Current file position is 8386
GPL Ghostscript 9.50: Unrecoverable error, exit code 1
=================================
gs -q -g156x76 -r384x384 -sDEVICE=ppmraw -sOutputFile=_form1.pnm -dNOPAUSE -dBATCH -dNOSAFER _form1.ps
=================================
```

**Comment by albert-github** [comment]
This was what I was afraid of it is not the `gs` command when it is invoked but the `ps2epsi ` which invokes the `gs` command.
On *nux it look to me that `ps2epsi` is a script file on `/bin` (at least at Cygwin it is), can you send me this `ps2epsi` file?

(On Cygwin there is no 9.50 yet, latest is 9.27. On windows the `ps2epsi` is an executable and comes with MikTeX).

**Comment by piyushrpt** [comment]
Here is the ps2epsi file on OSX via MacPorts - this is a script file. 
[ps2epsi.txt](https://github.com/doxygen/doxygen/files/4031798/ps2epsi.txt)

**Comment by albert-github** [comment]
The file is the same as file I found on Cygwin (not very strange as it is a script file).

In the file there are 2 places where `$GS_EXECUTABLE` is really used, i.e. line 52 and 94 (actually the first one is the important one, as far as I can see). Can you add here as a test the option `-dNOSAFER`?

**Comment by piyushrpt** [comment]
Turns out it was line 94 - Adding the option to the first call had no impact. With that option on line 94, the images are generated fine.

**Comment by albert-github** [comment]
That is good news.
Problem lies outside doxygen's scope, so it is a bit hard to fix for us (as far as I can see no optional arguments can be specified with `ps2epsi`), maybe best is to create an issue at https://bugs.ghostscript.com/

@mahehr 
- can you run with the 1.8.17 version?

**Comment by albert-github** [comment]
@marehr @malaterre 

I'm not sure whether or not edits in comments are noted, so I added this just for the notification purpose.

**Comment by IsaacPascual** [comment]
My ps2epsi file is like @piyushrpt one

**Comment by IsaacPascual** [comment]
I try `-dNOSAFER` and work but not sure if the correct. But not sure if it is considered really a bug from ghostscript project. There is no way to put `--permit-file-read=_form0.eps` from doxygen?

**Comment by albert-github** [comment]
@IsaacPascual I don't get the remark: 
>There is no way to put `--permit-file-read=_form0.eps` from doxygen?

when `ps2epsi` would accept an option or something similar like an environment variable it would, probably, be possible to make the change in doxygen, but I don't see such a possibility in the `ps2epsi` code

**Comment by albert-github** [comment]
I've just created an issue with ghostscpt: https://bugs.ghostscript.com/show_bug.cgi?id=702024

**Comment by IsaacPascual** [comment]
Well! We see what happen on the other side! I think that a way to bridge some options on `ps2epsi`'s `gs` command will be very good.

**Comment by IsaacPascual** [comment]
How about use new `eps2eps` script? It look simpler than `ps2epsi` but I don't really know if this is all that `doxygen` needs. My distro `ghostscript` package install it on `/usr/bin/`.

**Comment by albert-github** [comment]
@IsaacPascual 
Indeed the `eps2eps` script look very simple. I have some small doubts because of the `-dSAFER` although my tests didn't reveal a problem (older version).
- Can you give it a try with the 9.0 version, i.e. in my original scrip just replace `ps2epsi` by `eps2eps`.

**Comment by marehr** [comment]
> @marehr @malaterre
> 
> I'm not sure whether or not edits in comments are noted, so I added this just for the notification purpose.

I got the notification :) How can I help you?

**Comment by albert-github** [comment]
@marehr thanks for your reply.

In the issue #7290 you were one of the commenters that had a problem with formulas in doxygen in combination with gs9.50. 

**Possible help**:
- in comment https://github.com/doxygen/doxygen/issues/7484#issuecomment-571532453 I posted a forms.tar,gz that simulates what doxygen 1.8.17 does, maybe you can give this a try with gs9.50, this should give similar problems as OP found
- replace in the `cmds`  script the (all) `ps2epsi` by `eps2eps`, now there should not be any problems.

When I have confirmation that the `eps2eps` works I can use this in a proposed patch for doxygen, otherwise I have to see whether I can find another solution.

__Some background__:
In this (#7484) issue a new problem surfaces regarding gs9.50.
In #7290 we concentrated on the call to `gs` where the `-dNOSAFER` was added to work with gs9.50.
In the mean time also the statement regarding `ps2epsi` had been added and this script (!) uses also `gs` and looks like to suffer from similar problems.
In comment https://github.com/doxygen/doxygen/issues/7484#issuecomment-572019884 I mentioned an issue created with ghostscript and here it was "advised" to use `eps2eps` instead if `ps2epsi`. My forts test look OK but I don't have gs9.50 to test it.

**Comment by IsaacPascual** [comment]
Hi! I subtitute `ps2epsi` by symbolic link to `eps2eps` (just because I can't recompile all doxygen) and all works fine!
There is only one question... There is a verbose message (one for each formula I think) that say:
```DEBUG: FC_WEIGHT didn't match```

**Comment by albert-github** [comment]
Thanks for the answer.

Looks like this message comes as explained in https://tug.org/pipermail/texhax/2019-March/023542.html

As you see everything is correct I think we can ignore this message.

**Comment by albert-github** [comment]
I've just pushed a proposed patch, pull request #7496 

(@marehr confirmation as requested in https://github.com/doxygen/doxygen/issues/7484#issuecomment-572503569 still won't hurt).

**Comment by isaac-pascual** [comment]
There is a hundreds of this message in my use case... I prefer that work but all this verbose messages are not so good.

**Comment by albert-github** [comment]
I agree that these messages would be quite inconvenient.
This problem lies, as far as I can see,  outside of doxygen's scope ...
Which command does give this message? (just try with the test script).
Maybe we can create an issue with ghostscript.

**Comment by albert-github** [comment]
Code has been integrated in master on github (please don't close the issue as this will be done at the moment of a release).

**Comment by doxygen** [comment]
@albert-github The change from `dvips` to `eps2eps` causes regression it seems (at least on my system). The spacing around formulas (bounding box) is no longer correct:

Compare the current master with too much spacing:
<img width="311" alt="Screenshot 2020-02-09 at 17 23 16" src="https://user-images.githubusercontent.com/1300762/74105931-5600e980-4b62-11ea-90a1-5234b21828af.png">


against 1.8.7 with the right amount of spacing.
<img width="321" alt="Screenshot 2020-02-09 at 17 23 53" src="https://user-images.githubusercontent.com/1300762/74105835-a1ff5e80-4b61-11ea-9a32-6fb1c83fd0dd.png">

Input is a file `test.cpp` with a default config file:
~~~{c}
/*! \mainpage
 *
 *  This is \f$x\f$ and \f$y\f$.
 */
~~~

**Comment by albert-github** [comment]
@doxygen 

What is the used version of `gs` / `gswin32c`:
  - 1.8.17
  - master

 I have to investigate deeper, but first thought is that it is dependent on the ghostscript version.
With different versions (1.8.16 / 1.8.17 / master) and different ghostscript versions I get different results (Maybe another argument for using MathJax :wink:).

Can you open a new issue for this problem (the current issue is due to the fact that the formulas are not shown at all) ?

**Comment by albert-github** [comment]
@doxygen 

Forgot to request the html directories of the both versions or at least the `form*` files from these directories (so we can compare the files).

**Comment by doxygen** [comment]
@albert-github Done, see https://github.com/doxygen/doxygen/issues/7578

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.8.18.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
